### ISSUE TITLE
DataLoader with option to re-use worker processes

### ISSUE THREAD

**Issue by dashesy** [pytorch/pytorch#15849]
## 🚀 Feature
Currently after an epoch is ended dataloader spawns a new process to read data. This means if processes have cached some internal state (db connection, indexing, ...) they will be lost and the new process will have the overhead of creating the connection or indexing.

## Motivation

I have a custom dataset (that indexes data for fast retrieval) and only part of the dataset is used in training. If I have a training with longer epochs, training will be faster than shorter epochs because when an epoch ends a new process will have to re-read and index the custom dataset.

## Pitch

A flag to tell dataloader to re-use the worker processes. If the process pool has the re-use option ([Loky from joblib](https://github.com/tomMoral/loky) does) that would be enough.

## Alternatives

Documentation should specify this issue. Option to use [Loky](https://github.com/tomMoral/loky) for all multi-processing backend will be also great.

## Additional context

Started a discussion [here](https://discuss.pytorch.org/t/dataloader-re-use-worker-processes/34071)

**Comment by ezyang** [comment]
This seems like a good idea, it will also help solve performance problems on Windows where process spawning is expensive, see #12831

**Comment by SsnL** [comment]
FWIW, this didn't bring much speed-up on linux when I tried it ~2 months ago.

**Comment by SsnL** [comment]
@peterjc123 nah... I don't think so. I can try coding up one for the new refactored data loader this weekend. :)

**Comment by peterjc123** [comment]
@SsnL Any news on this?

**Comment by ekmungi** [comment]
Did anyone encounter a similar problem under linux ? I am using Pytorch 1.0.0 and with my threadripper, I found this issue slowing down the dataloader almost 4 times, shown here: https://github.com/pytorch/pytorch/issues/12831#issuecomment-455277786

In my case, I created a custom dataset and just wrap it with the pytorch DataLoader.

The data is basically a collection of videos. I randomly choose a video and from that choose a random frame. To do that I create a dataset for each video and combine all of them using ConcatDataset.

**Comment by SsnL** [comment]
@peterjc123 sorry... I didn’t get to work on this. I’ll probably do it next week.

**Comment by peterjc123** [comment]
@Ssnl No worries. Please do it when you have time.

**Comment by kg1980** [comment]
Hi, any updates regarding this Issue?

**Comment by TimZaman** [comment]
@SsnL Great! Another good reason for this is that this would prohibit the usage of any caching in the dataloader. For example, some hocus pocus with labels or e.g. polygon/bounding-box rasterization. I could make an LRU cache, but it'd just be trashed with every epoch.

**Comment by lizhitwo** [comment]
An addition advantage is if I have a very short epoch (~10 iterations) but a large batch size, the initialization of dataloader iterator would take a huge chunk of running time of my epoch for every epoch. Sometimes my congested machine would even run into issues creating these processes, making the initialization as slow as 30 seconds. Being able to reuse them would cut this time down.

**Comment by PetrochukM** [comment]
This is super easy to implement folks:
```python
import torch
import torch.utils.data


class _RepeatSampler(object):
    """ Sampler that repeats forever.

    Args:
        sampler (Sampler)
    """

    def __init__(self, sampler):
        self.sampler = sampler

    def __iter__(self):
        while True:
            yield from iter(self.sampler)

    def __len__(self):
        return len(self.sampler)


class DataLoader(torch.utils.data.dataloader.DataLoader):

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.batch_sampler = _RepeatSampler(self.batch_sampler)
        self.iterator = super().__iter__()

    def __iter__(self):
        for i in range(len(self)):
            yield next(self.iterator)
```

**Comment by ezyang** [comment]
@zhangguanheng66 this issue did not have enough triage labels on it to be considered triaged. Could you please retriage it?

**Comment by lizhitwo** [comment]
> 
> 
> This is easy to implement folks, I recommend using this:
> 
> ```python
> import torch
> import torch.utils.data
> 
> 
> class _RepeatSampler(object):
>     """ Sampler that repeats forever.
> 
>     Args:
>         sampler (Sampler)
>     """
> 
>     def __init__(self, sampler):
>         self.sampler = sampler
> 
>     def __iter__(self):
>         while True:
>             yield from iter(self.sampler)
> 
>     def __len__(self):
>         return len(self.sampler)
> 
> 
> class DataLoader(torch.utils.data.dataloader.DataLoader):
> 
>     def __init__(self, *args, **kwargs):
>         super().__init__(*args, **kwargs)
>         self.batch_sampler = _RepeatSampler(self.batch_sampler)
>         self.iterator = super().__iter__()
> 
>     def __iter__(self):
>         for i in range(len(self)):
>             yield next(self.iterator)
> ```
> 
> @SsnL Let me know if you'd expect this to break anything. Initially, it looks like the `DataLoader` will continue to load data even if it's not being processed. Are there any limits on the maximum size of the `data_queue`?

This does not seem to address the issue. This issue is about the object `iter(self.sampler)` inside this `_RepeatSampler.__iter__()` being reused, not reconstructed each epoch.

**Comment by PetrochukM** [comment]
@lizhitwo No, it's not if your worried about "worker processes".

The `batch_sampler` default implementation, if you didn't override it with the optional parameter, does not do much processing at all or have any workers. The private object `_DataLoaderIter` owns the workers. I construct it once via: `self.iterator = super().__iter__()`.

**Comment by lizhitwo** [comment]
Oh sorry my bad. I thought you were inheriting the dataloader iterator.

**Comment by SsnL** [comment]
@PetrochukM This is elegant! That will mostly work and cover probably 90% use cases! To get the same functionality in core, there are some extra constraints through:
1. can not depend on `len(loader)` as it depends on `len(dataset)` which is implemented by user and may not be accurate
2. need to work with `IterableDataset`
3. can not overwrite `batch_sampler` as user code may depend on it. although it is not a documented attr, but it is not underscored and I think I've heard people using it...

**Comment by PetrochukM** [comment]
@SsnL Your right. I don't see this making it into the core. That said, I see this as covering most use cases.

I'm still worried about `data_queue`. If the `DataLoader` is loading faster than batches are consumed, will the `data_queue` continue to grow in memory (i.e. now that the sampler has an infinite length).

**Comment by SsnL** [comment]
@PetrochukM I agree that this covers most use cases. 

The maximum number of outstanding batches is upper limited by `2 * num_workers` so that should not be a big issue.

**Comment by PetrochukM** [comment]
@SsnL Awesome! I saw an `assert` along those lines but I didn't quite understand it, thanks for the confirmation.

**Comment by PetrochukM** [comment]
I ran the above code for a multiday training session and it worked well! I had a 5% - 10% improvement in speed across two separate models.

**Comment by Masseeh** [comment]
> This is easy to implement folks, I recommend using this:
> 
> ```python
> import torch
> import torch.utils.data
> 
> 
> class _RepeatSampler(object):
>     """ Sampler that repeats forever.
> 
>     Args:
>         sampler (Sampler)
>     """
> 
>     def __init__(self, sampler):
>         self.sampler = sampler
> 
>     def __iter__(self):
>         while True:
>             yield from iter(self.sampler)
> 
> 
> class DataLoader(torch.utils.data.dataloader.DataLoader):
> 
>     def __init__(self, *args, **kwargs):
>         super().__init__(*args, **kwargs)
>         self.batch_sampler = _RepeatSampler(self.batch_sampler)
>         self.iterator = super().__iter__()
> 
>     def __len__(self):
>         return len(self.batch_sampler.sampler)
> 
>     def __iter__(self):
>         for i in range(len(self)):
>             yield next(self.iterator)
> ```
> 
> @SsnL Let me know if you'd expect this to break anything. Initially, it looks like the `DataLoader` will continue to load data even if it's not being processed. Are there any limits on the maximum size of the `data_queue`?

By replacing `self.iterator = super().__iter__()` with `self.iterator = None`  in the constructor and initializing `self.iterator` in the `__iter__` function for the first time and then reusing it, we can reduce the memory footprint a little bit further.

**Comment by ezyang** [comment]
Tentatively bumping up priority due to activity

**Comment by cpuhrsch** [comment]
Flagging this for https://github.com/pytorch/pytorch/issues/24915

**Comment by jonmorton** [comment]
The solution suggested by @PetrochukM no longer works with the latest pytorch due to the `batch_sampler` being made private. Even getting around that, it does not work. 

The solution that worked for me was to lie about the length of my dataset, and then use modulo to find the real item:

```
class MyDataset(Dataset):
    def __init__(self, items):
        self._real_len = len(items)
    def __len__(self):
        return self._real_len * 50000
    def __getitem__(self, index):
        index = index % self._real_len
        ...
```

**Comment by PetrochukM** [comment]
@jonmorton I asked about the private changes here: https://github.com/pytorch/pytorch/pull/23744

It's preventing me more fully upgrading to PyTorch 1.3 :( 

Could you use `Math.inf` for `__len__` instead of a really large number? It might make it a bit cleaner.

**Comment by odedbd** [comment]
This is blocking my upgrade to 1.3 as well. I have a use case where setting up the workers is a costly process.

**Comment by cpuhrsch** [comment]
@SsnL - can I assign this to you? I'll assign to me just to keep track.

**Comment by SsnL** [comment]
@cpuhrsch - Sure! Expect no progress before winter holidays though.

**Comment by SsnL** [comment]
The issue can be worked around by invoking `object.__setattr__`. @odedbd

**Comment by odedbd** [comment]
@SsnL thank you for the pointer, that worked great!

**Comment by sampepose** [comment]
What's the latest status on this PR? This'll be useful when forking is slow like with forkserver.

**Comment by rmcavoy** [comment]
Does this code still work taking into account the change to generators in 3.7 i.e. they convert a stop iteration exception into an error? If someone has a version working in python 3.7 and/or PyTorch 1.3 it would be much appreciated if they could post it even if it still has the restrictions discussed above.

**Comment by ezyang** [comment]
@rmcavoy please post a new issue for your problem

**Comment by rmcavoy** [comment]
Why? I am talking about the code posted above is this issue alone. As far as I am aware, the only place that infinite generator code exists is in the comments of this issue.

**Comment by ezyang** [comment]
Oops, my bad! I thought you were referring to a different generator problem

**Comment by mranzinger** [comment]
Based on @Masseeh 's recommendation, this is what I'm using in Py 3.7, Torch 1.3:
```Python
import torch
import torch.utils.data


class _RepeatSampler(object):
    """ Sampler that repeats forever.

    Args:
        sampler (Sampler)
    """

    def __init__(self, sampler):
        self.sampler = sampler

    def __iter__(self):
        while True:
            yield from iter(self.sampler)


class FastDataLoader(torch.utils.data.dataloader.DataLoader):

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        object.__setattr__(self, 'batch_sampler', _RepeatSampler(self.batch_sampler))
        self.iterator = super().__iter__()

    def __len__(self):
        return len(self.batch_sampler.sampler)

    def __iter__(self):
        for i in range(len(self)):
            yield next(self.iterator)
```
Using it just like the vanilla loader. My life is 2x happier now, though YMMV. Hope this helps.

**Comment by rmcavoy** [comment]
Thanks! Apparently the actual problem was I had gotten the syntax for setting the batch sampler wrong so the iterator was finishing which is why it gave the stop iteration error was even being raised.

**Comment by s-rog** [comment]
> 
> 
> Based on @Masseeh 's recommendation, this is what I'm using in Py 3.7, Torch 1.3:
> 
> ```python
> import torch
> import torch.utils.data
> 
> 
> class _RepeatSampler(object):
>     """ Sampler that repeats forever.
> 
>     Args:
>         sampler (Sampler)
>     """
> 
>     def __init__(self, sampler):
>         self.sampler = sampler
> 
>     def __iter__(self):
>         while True:
>             yield from iter(self.sampler)
> 
> 
> class FastDataLoader(torch.utils.data.dataloader.DataLoader):
> 
>     def __init__(self, *args, **kwargs):
>         super().__init__(*args, **kwargs)
>         object.__setattr__(self, 'batch_sampler', _RepeatSampler(self.batch_sampler))
>         self.iterator = super().__iter__()
> 
>     def __len__(self):
>         return len(self.batch_sampler.sampler)
> 
>     def __iter__(self):
>         for i in range(len(self)):
>             yield next(self.iterator)
> ```
> 
> Using it just like the vanilla loader. My life is 2x happier now, though YMMV. Hope this helps.

Can this be applied to distributed sampler as is?

**Comment by pisarik** [comment]
Does workaround proposed above by @mranzinger work?

**Test code:**
```python
import os
from torch.utils import data


class DummyDataset(data.Dataset):
    def __len__(self):
        return 10
    def __getitem__(self, item):
        return item


for epoch in range(2):
    print('Epoch: ', epoch)
    for batch in FastDataLoader(DummyDataset(), num_workers=2, batch_size=5,
                                worker_init_fn=lambda x: print('pid: ', os.getpid())):
        print(batch)
```

**Output:**
```
Epoch:  0
pid:  9406
pid:  9407
tensor([0, 1, 2, 3, 4])
tensor([5, 6, 7, 8, 9])
Epoch:  1
pid:  9414
pid:  9415
tensor([0, 1, 2, 3, 4])
tensor([5, 6, 7, 8, 9])
```

Every epoch `worker_init_fn` is called and pid differs for each epoch.

**Comment by skyoung** [comment]
Strangely, I found that the memory usage on my first GPU is much higher than other gpus when I use this data loader for multi-gpu training.

**Comment by zhangguanheng66** [comment]
> Strangely, I found that the memory usage on my first GPU is much higher than other gpus when I use this data loader for multi-gpu training.

@skyoung Are you using `DataParallel`?

**Comment by skyoung** [comment]
> > Strangely, I found that the memory usage on my first GPU is much higher than other gpus when I use this data loader for multi-gpu training.
> 
> @skyoung Are you using `DataParallel`? If so, this is expected. `DDP` will give you a more balanced gpu usage.

No. I use `DDP`. When I remove this data loader and use the normal one, then the memory usage imbalance disappears. It seems the problem is caused by `__iter__()` of `FastDataLoader`. If I delete it, everything goes well. but it will not raise stop error at the end of the epoch.

**Comment by zhangguanheng66** [comment]
cc @mrshenli

**Comment by skyoung** [comment]
Setting `pin_memory=False` solves the GPU memory problem, but slows down the data loading speed. I guess `FastDataLoader` is not compatible with the pin memory mechanism.

**Comment by Red-Eyed** [comment]
Although @Masseeh suggestion works faster on Windows, it uses too much RAM, and sometimes it crashes due to OOM :( 
Had to switch back on vanilla DataLoader

**Comment by AdrianKs** [comment]
Using the `FastDataloader` as described here, seems to have a bad influence on my resulting model quality. The only difference I can see to the standard dataloader, which re-creates the workers after every epoch, is the re-seeding taking place on creation of the workers. Re-seeding manually does not seem to help with the reduction in quality.

Is there anything else that could influence the training procedure by using the `FastDataloader`?

**Comment by pisarik** [comment]
I recently saw that `persistent_worker` parameter appeared in [DataLoader](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader) in pytorch 1.8.0

> persistent_workers (bool, optional) – If True, the data loader will not shutdown the worker processes after a dataset has been consumed once. This allows to maintain the workers Dataset instances alive. (default: False)

So shouldn't be this issue closed at all?

**Comment by s-rog** [comment]
https://github.com/pytorch/pytorch/commit/5472426b9f85c8107aade5d256cf4cde572eab5c Seems to be present since 1.7.0

**Comment by AdrianKs** [comment]
Thanks, I've totally overlooked this new implementation.

**Comment by mszhanyi** [comment]
One more test.  For same script, change num-workers with spawn and  persistent_workers =True
num_workers | Windows | Linux
-- | -- | --
2 | 45.46 | 36.43
4 | 22.96 | 23.78
8 | 12.55 | 23.64

**Comment by AdrianKs** [comment]
@xx-zhou16 Are you at any point changing something in the dataset? As the dataloader-worker is kept alive in a separate process, the changes to the dataset won't propagate to the workers.

**Comment by xx-zhou16** [comment]
@AdrianKs https://github.com/pytorch/pytorch/issues/15849#issuecomment-838188281 https://github.com/pytorch/pytorch/issues/15849#issuecomment-838142196
Nothing is changed in the dataset during the training. 

And I also find that, if I evaluate a normally trained model (only inference, but not training), the metrics of the predictions are not correct (extremely low). In detail, I do not `FastDataLoader` during training and use  `FastDataLoader` only for inference. And the model infers over the same validation set several times (epochs). **And only the result of the first epoch** is different from that without ``FastDataLoader`.

**Comment by xx-zhou16** [comment]
https://github.com/pytorch/pytorch/issues/15849#issuecomment-839192912 @ezyang
I don't think it is about RNG.
1. For training with `FastDataLoader`, if it is only different on RNG, the models should be similar on accuracy. But now the models trained with and without `FastDataLoader` are so different on accuracy.   
2. For inference, it is definitely not related to RNG. No random is used there. The evaluation of the **same** trained model is so different on the first epoch of the validation set. But the other epochs seem the same.

**Comment by xx-zhou16** [comment]
https://github.com/pytorch/pytorch/issues/15849#issuecomment-839192912 @ezyang
I don't think it is about RNG.
1. For training with `FastDataLoader`, if it is only different on RNG, the models should be similar on accuracy. But now the models trained with and without `FastDataLoader` are so different on accuracy.   
2. For inference, it is definitely not related to RNG. No random is used there. The evaluation of the **same** trained model is so different on the first epoch of the validation set. But the other epochs seem the same.

**Comment by xx-zhou16** [comment]
https://github.com/pytorch/pytorch/issues/15849#issuecomment-839192912 @ezyang
I don't think it is about RNG.
1. For training with `FastDataLoader`, if it is only different on RNG, the models should be similar on accuracy. But now the models trained with and without `FastDataLoader` are so different on accuracy.   
2. For inference, it is definitely not related to RNG. No random is used there. The evaluation of the **same** trained model is so different on the first epoch of the validation set. But the other epochs seem the same.

**Comment by xx-zhou16** [comment]
There is another thing to be noted:   
1. use `FastDataLoader` leads to much lower accuracy (very apparently at the beginning of training). But it can speed up the training procedure. 
2. But everything is alright if I use `_persistent_worker=True` on Pytorch Official Implementation. The training and inference are both normal. And it can also speed up.  

But not all my experiments can be done on Pytorch 1.7 or 1.8. Because I have very limited amount of RTX 3060 GPUs, and also have some other versions of GPU (such as Xp and K40). Cuda11 is not compatible with the latter. So I still need to figure out the essential problem.

**Comment by Wuziyi616** [comment]
When using `FastDataLoader` with PyTorch 1.10, my code gets stuck when creating dataloader. So I have to switch to vanilla `DataLoader` with `persistent_worker=True`

**Comment by willyd** [comment]
@xx-zhou16 I could be wrong but the loss in accuracy could be due to the fact that the data does not get reshuffled at evey epoch when using FastDataLoader? I noticed something like this in the YOLOv5 implementation: https://github.com/ultralytics/yolov5/issues/6741.
